### PR TITLE
Refactor energy loss fluctuations

### PR DESCRIPTION
**Pull request by sethrj**
The only substantive change here is that `calc_energy_loss` resamples the energy loss fluctuation if the distribution is greater or equal to the pre-step energy, rather than clipping to the energy (and causing the particle to end a step at zero energy even when the range wasn't the step limiter).

To help investigate the energy loss fluctuation, I broke the `EnergyLossDistribution` into distinct models for the energy loss: `Delta` (low energy or low mean loss), `Urban` (thin layers), `Gaussian` (heavy particles), and `Gamma` (heavy particles with small loss). These classes are now "detail" classes returned via a traits class by `make_distribution` in `EnergyLossDistribution`. Each class can be individually constructed with its input parameters for more detailed testing, or built directly from the `EnergyLossHelper` class that's used to decide the appropriate model.

Because the distributions calculate many properties in their constructors (previously calculated inside `operator()`) they are more efficient when resampling. After ensuring that the unit tests produced exactly the same answers when reconstructing the distributions inside the loop, I moved the constructor outside, which saves many calls to the RNG for the gaussian distribution (because it's able to make use of the second normal variable it samples for the "even" calls to `operator()`).

I also made a couple of tweaks to the Urban code, replacing manual calls to `generate_canonical` with appropriate distributions.

The static polymorphism dispatch (switch + macro) is rather clumsy; we should look into replacing with a recursive template call or other alternatives. I just selected this method because it's easy to visually parse.

I ran the "medium" input on CPU with debug assertions enabled: the error messages have been eliminated, and no assertions are raised.

See #354.

### CODE REVIEW

**Review by amandalund — COMMENTED** [review]
Nice @sethrj, thanks for going through this!

File: src/physics/base/PhysicsStepUtils.i.hh [hunk 1]
```diff
@@ -144,74 +169,91 @@ CELER_FUNCTION ParticleTrackView::Energy
...
+        // without going through the condition above.
+        auto calc_energy
+            = physics.make_calculator<InverseRangeCalculator>(grid_id);
+        eloss = pre_step_energy - value_as<Energy>(calc_energy(range - step));
     }
 
-    // Add energy loss fluctuations if this is the "energy loss" process
-    if (eloss > 0 && eloss < pre_step_energy.value()
-        && physics.add_fluctuation())
+    if (physics.add_fluctuation() && eloss > 0 && eloss < pre_step_energy)
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/base/PhysicsStepUtils.i.hh:230]: I had the same thought -- which is why the last commit is there, because that led to an infinite loop. What was happening is that `range - step` was someting like 1e-30, which meant that `eloss = pre_step_energy`. I think we can change it to an assertion once we have different "step limiters", but for now it has to be in this conditional.

File: src/physics/em/detail/EnergyLossUrbanDistribution.hh [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    loss_scaling_
+        = max_energy_ > this->fwhm_min_energy()
+              ? 1 + real_type(0.5) * this->fwhm_min_energy() / max_energy_
+              : real_type(1.5);
+#if 0
+    // Alternatively...
+    loss_scaling_ = real_type(0.5)
+                        * std::fmin(this->fwhm_min_energy() / max_energy_, 1)
+                    + real_type(1);
+#endif
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/em/detail/EnergyLossUrbanDistribution.hh:161]: Thanks for reminding me. Let me actually write a "cuda snippets" test case here just to see the difference: the first case is what's literally written in geant, to "conditionally" save the division, but the second requires no branching and thus might actually be faster on both GPU/CPU. And I'm not so much concerned about performance as clarity: the second I think shows the bounds more clearly, but I wanted to leave the original in there till the unit tests were complete.

[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/em/detail/EnergyLossUrbanDistribution.hh]: So CUDA is "faster": intrinsic "min" and no branching (see https://github.com/celeritas-project/cuda-test-snippets/tree/master/energy-loss-urban). CPU is "slower": external call to `_fmin` or unconditional evaluation of the divisor. I've added fmin(/min/umin)  overloads for CUDA device code, so that CPU will at least inline the "min" function, and optimize for CUDA performance by using its built-ins. This way I don't have to feel bad about substantial CPU performance hits, the GPU version is faster, and most importantly the expression ends up more readable.

File: src/physics/base/PhysicsStepUtils.i.hh [hunk 3]
```diff
@@ -23,6 +23,28 @@
...
+{
+    using Energy = EnergyLossHelper::Energy;
+    Energy result;
+    auto   sample_eloss = make_distribution<M>(helper);
+
+    do
+    {
+        result = sample_eloss(rng);
+        // Resample if the energy loss is above the particle's energy
+    } while (value_as<Energy>(result) >= value_as<Energy>(max_loss));
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/base/PhysicsStepUtils.i.hh:43]: Yeah, maybe I should just suspend this "resampling" change for now? I did this refactor to make it easier to dive into what each of the distributions look like for the misbehaving tracks.

File: src/physics/em/detail/EnergyLossUrbanDistribution.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/em/detail/EnergyLossUrbanDistribution.hh]: Excellent!

File: src/physics/base/PhysicsStepUtils.i.hh (hunk 3 above)
[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/base/PhysicsStepUtils.i.hh:43]: Yeah, I'd lean toward holding off on this until we dig into it more (which will be much easier now!)

[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/base/PhysicsStepUtils.i.hh]: OK done.

**Review by amandalund — APPROVED** [review]
Thanks @sethrj!

**Comment by sethrj** [comment]
Thank you @amandalund for looking over this large changeset!
